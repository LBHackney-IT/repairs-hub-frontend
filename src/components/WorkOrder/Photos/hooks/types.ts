export type Link = {
  presignedUrl: string
  key: string
}
