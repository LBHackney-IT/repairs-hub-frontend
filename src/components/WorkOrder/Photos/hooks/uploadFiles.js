import axios from 'axios'
import imageCompression from 'browser-image-compression'
import { frontEndApiRequest } from '@/root/src/utils/frontEndApiClient/requests'

const uploadFiles = async (
  files,
  workOrderReference,
  uploadGroupLabel,
  description,
  setLoadingStatus = null
) => {
  if (setLoadingStatus !== null) {
    setLoadingStatus(`Uploading ${files.length} file(s)`)
  }

  // 1. get presigned urls
  const uploadUrlsResult = await getUploadLinks(
    workOrderReference,
    files.length
  )

  if (!uploadUrlsResult.success) {
    if (setLoadingStatus !== null) setLoadingStatus(null)

    return {
      success: false,
      requestError: uploadUrlsResult.error,
    }
  }

  const presignedUrls = uploadUrlsResult.result.links

  function onProgress(completed, total) {
    setLoadingStatus(`${completed} of ${total} file(s) uploaded`)
  }

  // 2. Upload files to S3
  const uploadFilesToS3Response = await uploadFilesToS3(
    files,
    presignedUrls,
    onProgress
  )

  if (!uploadFilesToS3Response.success) {
    if (setLoadingStatus !== null) setLoadingStatus(null)

    return {
      success: false,
      requestError: uploadFilesToS3Response.error,
    }
  }

  // 3. Complete upload
  const completeUploadResult = await completeUpload(
    workOrderReference,
    presignedUrls.map((x) => x.key),
    uploadGroupLabel,
    description
  )

  if (!completeUploadResult.success) {
    if (setLoadingStatus !== null) setLoadingStatus(null)

    return {
      success: false,
      requestError: completeUploadResult.error,
    }
  }

  return {
    success: true,
  }
}

const completeUpload = async (
  workOrderReference,
  s3Keys,
  uploadGroupLabel,
  description
) => {
  try {
    const result = await frontEndApiRequest({
      method: 'post',
      path: `/api/workOrders/images/completeUpload`,
      requestData: {
        workOrderReference: workOrderReference,
        s3Objects: s3Keys,
        uploadGroupLabel,
        description,
      },
    })

    return { success: true, result }
  } catch (error) {
    return { success: false, error }
  }
}

const uploadFilesToS3 = async (files, links, onProgress) => {
  const promiseList = []
  let completed = 0

  files.forEach((file, i) => {
    promiseList.push(uploadFileToS3(file, links[i]))
  })

  // wrap with callback to track how many have been completed
  const wrappedPromises = promiseList.map((promise) =>
    promise.then((result) => {
      completed += 1
      onProgress(completed, promiseList.length)
      return result
    })
  )

  try {
    const result = await Promise.allSettled(wrappedPromises)

    return { success: true, result }
  } catch (error) {
    return { success: false, error }
  }
}

const uploadFileToS3 = async (file, link) => {
  const compressionOptions = {
    maxSizeMB: 1,
    maxWidthOrHeight: 1920,
    useWebWorker: true,
    maxIteration: 1,
  }

  const compressedFile = await imageCompression(file, compressionOptions)

  return axios.request({
    method: 'PUT',
    url: link.presignedUrl,
    data: compressedFile,
  })
}

const getUploadLinks = async (workOrderReference, numberOfFiles) => {
  try {
    const result = await frontEndApiRequest({
      method: 'get',
      path: `/api/workOrders/images/uploadLink`,
      params: {
        workOrderReference: workOrderReference,
        numberOfFiles: numberOfFiles,
      },
    })

    return { success: true, result }
  } catch (error) {
    return { success: false, error }
  }
}

export default uploadFiles
