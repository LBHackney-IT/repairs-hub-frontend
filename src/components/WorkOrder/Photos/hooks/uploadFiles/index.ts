import getPresignedUrls from './getPresignedUrls'
import uploadFilesToS3 from './uploadFilesToS3'
import completeUpload from './completeUpload'
import { captureException } from '@sentry/nextjs'
import fileUploadStatusLogger from './fileUploadStatusLogger'
import compressFile from './compressFile'
import { getCachedFile, setCachedFile } from './cacheFile'
import ensureAllFilesReadable from './ensureAllFilesReadable'
import useCloudwatchLogger from '@/root/src/utils/cloudwatchLogger'

export class FileUploadError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'FileUploadError'
  }
}

const uploadFiles = async (
  files: File[],
  workOrderReference: string,
  description: string,
  uploadGroupLabel: string,
  setUploadStatus: (status: string) => void,
  cwLogger: ReturnType<typeof useCloudwatchLogger>
): Promise<{
  success: boolean
  requestError?: string
}> => {
  try {
    const statusLogger = fileUploadStatusLogger(files.length, setUploadStatus)

    // 1. Get cached files to upload or compress and cache them
    const filesToUpload: File[] = []
    const compressionErrors: Error[] = []

    for (const file of files) {
      try {
        let fileToUpload = await getCachedFile(file)

        if (fileToUpload == null) {
          fileToUpload = await compressFile(file)
          await setCachedFile(fileToUpload)
        }

        filesToUpload.push(fileToUpload)
      } catch (error) {
        filesToUpload.push(file)
        compressionErrors.push(error as Error)
      } finally {
        statusLogger('Compress')
      }
    }

    // 2. Get presigned urls
    const uploadUrlsResult = await getPresignedUrls(
      workOrderReference,
      filesToUpload.length
    )
    if (!uploadUrlsResult.success)
      throw new FileUploadError(uploadUrlsResult.error as string)

    const presignedUrls = uploadUrlsResult.result.links

    // 3. Upload files to S3
    await ensureAllFilesReadable(filesToUpload)
    const uploadFilesToS3Response = await uploadFilesToS3(
      filesToUpload,
      presignedUrls,
      () => statusLogger('Upload')
    )
    if (!uploadFilesToS3Response.success)
      throw new FileUploadError(
        (uploadFilesToS3Response.error as string) +
          (compressionErrors.length &&
            `| Compression error: ${compressionErrors?.[0]?.message}`)
      )

    // 4. Complete upload
    const completeUploadResult = await completeUpload(
      workOrderReference,
      presignedUrls.map((x) => x.key),
      description,
      uploadGroupLabel
    )
    if (!completeUploadResult.success)
      throw new FileUploadError(completeUploadResult.error as string)
  } catch (error) {
    if (error instanceof FileUploadError) {
      cwLogger.error(error.message)
      captureException('Failed to upload photos', {
        tags: {
          section: 'File upload',
        },
        extra: {
          workOrderReference,
          files: JSON.stringify(
            files.map((file) => ({
              name: file.name,
              size: file.size,
              type: file.type,
            }))
          ),
          message: error.message,
        },
      })
    }
    return {
      success: false,
      requestError: error.message || 'An error occurred while uploading files',
    }
  }

  return {
    success: true,
  }
}

export default uploadFiles
