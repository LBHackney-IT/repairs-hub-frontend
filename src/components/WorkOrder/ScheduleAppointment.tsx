import { useContext, useEffect, useRef } from 'react'
import { WorkOrder } from '../../models/workOrder'
import ScheduleDRSAppointmentLink from './ScheduleDRSAppointmentLink'
import ScheduleInternalAppointmentLink from './ScheduleInternalAppointmentLink'
import { buildDataFromScheduleAppointment } from '../../utils/hact/jobStatusUpdate/notesForm'
import { frontEndApiRequest } from '../../utils/frontEndApiClient/requests'
import UserContext from '../UserContext'

interface Props {
  hasExistingAppointment: boolean
  workOrder: WorkOrder
  schedulerSessionId: string
  //   openExternalLinkEventHandler: () => void
  externalAppointmentManagementUrl: string
  workOrderReference: string
  resetSchedulerSessionId: () => void
}

const ScheduleAppointment = (props: Props) => {
  const {
    hasExistingAppointment,
    workOrder,
    schedulerSessionId,
    // openExternalLinkEventHandler,
    externalAppointmentManagementUrl,
    workOrderReference,
    resetSchedulerSessionId,
  } = props

  const { user } = useContext(UserContext)

  const userOpenedWBM = useRef(false)
  const focusRef = useRef(null)

  const onWindowFocusCallback = () => {
    if (!userOpenedWBM.current) return

    // refresh page to trigger manual sync
    window.location.reload()
  }

  useEffect(() => {
    const ref = window.addEventListener('focus', onWindowFocusCallback, false)

    focusRef.current = ref

    return () => {
      window.removeEventListener(focusRef.current, onWindowFocusCallback, false)
    }
  }, [])

  const openExternalLinkEventHandler = async () => {
    userOpenedWBM.current = true

    const jobStatusUpdate = buildDataFromScheduleAppointment(
      workOrderReference.toString(),
      `${user.name} opened the DRS Web Booking Manager`
    )

    await frontEndApiRequest({
      method: 'post',
      path: `/api/jobStatusUpdate`,
      requestData: jobStatusUpdate,
    })

    setTimeout(() => {
      // sessionId is single use, so reset it
      resetSchedulerSessionId()
    }, 500)
  }

  if (!externalAppointmentManagementUrl) {
    return (
      <ScheduleInternalAppointmentLink
        workOrderReference={workOrderReference}
        hasExistingAppointment={hasExistingAppointment}
        appointmentIsToday={workOrder.appointmentIsToday()}
      />
    )
  }

  if (schedulerSessionId) {
    return (
      <ScheduleDRSAppointmentLink
        workOrder={workOrder}
        schedulerSessionId={schedulerSessionId}
        openLinkEventHandler={openExternalLinkEventHandler}
        hasExistingAppointment={hasExistingAppointment}
        appointmentIsToday={workOrder.appointmentIsToday()}
      />
    )
  } else {
    console.error('Scheduler Session ID does not exist')
  }
}

export default ScheduleAppointment
