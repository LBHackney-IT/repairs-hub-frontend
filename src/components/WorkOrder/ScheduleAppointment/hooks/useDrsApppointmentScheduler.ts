import { frontEndApiRequest } from '@/root/src/utils/frontEndApiClient/requests'
import { buildDataFromScheduleAppointment } from '@/root/src/utils/hact/jobStatusUpdate/notesForm'
import { useContext, useEffect, useRef, useState } from 'react'
import UserContext from '../../../UserContext'
import { getOrCreateSchedulerSessionId } from '../../../../utils/frontEndApiClient/users/schedulerSession'

export const useDrsAppointmentScheduler = (workOrderReference: string) => {
  const [schedulerSessionId, setSchedulerSessionId] = useState<string | null>(
    null
  )
  const [isLoading, setIsLoading] = useState<boolean>(true)
  const [error, setError] = useState<string | null>(null)

  const { user } = useContext(UserContext)
  const userOpenedWBM = useRef(false)

  const fetchSchedulerSessionId = async () => {
    setIsLoading(true)

    try {
      const sessionId = await getOrCreateSchedulerSessionId()
      setSchedulerSessionId(sessionId)
    } catch (err) {
      console.error(err)
      setError(
        `Something went wrong fetching DRS Session: ${
          err?.message || 'Unknown error'
        }`
      )
    } finally {
      setIsLoading(false)
    }
  }

  const handleExternalLinkOpen = async () => {
    userOpenedWBM.current = true

    const jobStatusUpdate = buildDataFromScheduleAppointment(
      workOrderReference.toString(),
      `${user.name} opened the DRS Web Booking Manager`
    )

    await frontEndApiRequest({
      method: 'post',
      path: '/api/jobStatusUpdate',
      requestData: jobStatusUpdate,
    })

    setTimeout(() => fetchSchedulerSessionId(), 500)
  }

  useEffect(() => {
    const handleWindowFocus = () => {
      if (userOpenedWBM.current) {
        window.location.reload()
      }
    }

    window.addEventListener('focus', handleWindowFocus)

    return () => {
      window.removeEventListener('focus', handleWindowFocus)
    }
  }, [])

  useEffect(() => {
    fetchSchedulerSessionId()
  }, [])

  return {
    schedulerSessionId,
    isLoading,
    error,
    handleExternalLinkOpen,
  }
}
