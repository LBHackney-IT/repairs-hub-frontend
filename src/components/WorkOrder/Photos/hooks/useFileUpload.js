import { useState } from 'react'
import uploadFiles from './uploadFiles'
import validateFileUpload from './validateFileUpload'

const useFileUpload = (workOrderReference, onSuccess) => {
  const [loadingStatus, setLoadingStatus] = useState(null)
  const [validationError, setValidationError] = useState(null)
  const [requestError, setRequestError] = useState(null)
  const [uploadSuccess, setUploadSuccess] = useState(null)

  const [files, setFiles] = useState([])

  const handleSubmit = async (e) => {
    e.preventDefault()

    if (loadingStatus) return

    // reset
    setUploadSuccess(null)
    setRequestError(null)
    setValidationError(null)

    const validationResult = validateFileUpload(files)
    setValidationError(validationResult)
    if (validationResult !== null) return

    const description = e.target.description.value

    const uploadResult = await uploadFiles(
      files,
      workOrderReference,
      '',
      'Uploaded directly to work order',
      description,
      setLoadingStatus
    )

    setLoadingStatus(null)

    if (!uploadResult.success) {
      setRequestError(uploadResult.requestError.message)
      return
    }

    setUploadSuccess(
      ` ${files.length} ${
        files.length === 1 ? 'photo has' : 'photos have'
      } been added to the workOrder`
    )
    setFiles([])
    onSuccess()
  }

  return {
    files,
    handleSubmit,
    uploadSuccess,
    setFiles,
    validationError,
    requestError,
    loadingStatus,
  }
}

export default useFileUpload
