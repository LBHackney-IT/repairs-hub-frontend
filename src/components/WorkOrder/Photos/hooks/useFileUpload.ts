import { useState } from 'react'
import uploadFiles from './uploadFiles'
import validateFileUpload from './validateFileUpload'

const useFileUpload = (workOrderReference: string, onSuccess: () => void) => {
  const [loadingStatus, setLoadingStatus] = useState(null)
  const [validationError, setValidationError] = useState(null)
  const [requestError, setRequestError] = useState(null)
  const [uploadSuccess, setUploadSuccess] = useState(null)

  const [files, setFiles] = useState<File[]>([])

  const handleSubmit = async (e) => {
    e.preventDefault()

    if (loadingStatus) return

    // reset
    setUploadSuccess(null)
    setRequestError(null)
    setValidationError(null)

    const validationResult = validateFileUpload(files)
    setValidationError(validationResult)
    if (validationResult !== null) return

    const description = e.target.description.value

    const uploadResult = await uploadFiles(
      files,
      workOrderReference,
      description,
      'Uploaded directly to work order',
      setLoadingStatus
    )

    setLoadingStatus(null)

    if (!uploadResult.success) {
      setRequestError(uploadResult.requestError)
      return
    }

    setUploadSuccess(
      ` ${files.length} ${
        files.length === 1 ? 'photo has' : 'photos have'
      } been added to the work order`
    )
    setFiles([])
    onSuccess()
  }

  return {
    files,
    handleSubmit,
    uploadSuccess,
    setFiles,
    validationError,
    requestError,
    loadingStatus,
  }
}

export default useFileUpload
