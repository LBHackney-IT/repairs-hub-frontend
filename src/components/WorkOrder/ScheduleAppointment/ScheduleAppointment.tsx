import { WorkOrder } from '../../../models/workOrder'
import ScheduleDRSAppointmentLink from '../ScheduleDRSAppointmentLink'
import ErrorMessage from '../../Errors/ErrorMessage'
import Spinner from '../../Spinner'
import { useDrsAppointmentScheduler } from './hooks/useDrsApppointmentScheduler'

interface Props {
  hasExistingAppointment: boolean
  workOrder: WorkOrder
  externalAppointmentManagementUrl: string
  workOrderReference: string
}

const ScheduleAppointment = ({
  hasExistingAppointment,
  workOrder,
  workOrderReference,
}: Props) => {
  const {
    schedulerSessionId,
    isLoading,
    error,
    handleExternalLinkOpen,
  } = useDrsAppointmentScheduler(workOrderReference)

  if (isLoading) {
    return (
      <span
        style={{
          color: '#64748b',
          fontSize: '12px',
          display: 'flex',
          alignItems: 'center',
          justifyContent: 'flex-start',
        }}
      >
        <Spinner width={20} height={20} />
        <span style={{ margin: '0 0 0 10px' }}>Fetching DRS session...</span>
      </span>
    )
  }

  if (error) {
    return <ErrorMessage label={error} />
  }

  return (
    <ScheduleDRSAppointmentLink
      openLinkEventHandler={handleExternalLinkOpen}
      schedulerSessionId={schedulerSessionId}
      hasExistingAppointment={hasExistingAppointment}
      appointmentIsToday={workOrder.appointmentIsToday()}
      workOrder={workOrder}
    />
  )
}

export default ScheduleAppointment
