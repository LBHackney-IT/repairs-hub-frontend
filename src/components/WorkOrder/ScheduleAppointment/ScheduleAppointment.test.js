jest.mock('./hooks/useDrsApppointmentScheduler', () => ({
  useDrsAppointmentScheduler: jest.fn(),
}))

import { useDrsAppointmentScheduler } from './hooks/useDrsApppointmentScheduler'

import { render } from '@testing-library/react'
import UserContext from '../../UserContext'
import { agent } from '@/root/factories/agent'
import { WorkOrder } from '@/root/src/models/workOrder'
import ScheduleAppointment from './ScheduleAppointment'

const workOrderData = {
  reference: 10000012,
  dateRaised: '2021-01-18T15:28:57.17811',
  lastUpdated: null,
  priority: 'A priority',
  property: '16 Pitcairn House  St Thomass Square',
  owner: 'HH General Building Repair - H01',
  description: 'This is a repair description',
  propertyReference: '00012345',
  status: 'In Progress',
  priorityCode: 0,
  raisedBy: 'Dummy Agent',
  target: '2021-01-23T18:30:00.00000',
  tradeCode: 'PL',
  tradeDescription: 'PLUMBING - PL',
  callerName: 'Jill Smith',
  callerNumber: '07700 900999',
  contractorReference: 'H01',
  appointment: null,
}

describe('AppointmentDetails component', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('Shows a loading state', () => {
    useDrsAppointmentScheduler.mockReturnValue({
      schedulerSessionId: 'schedulerSessionId',
      isLoading: true,
      error: null,
      handleExternalLinkOpen: jest.fn(),
    })

    let workOrder = new WorkOrder(workOrderData)

    const { asFragment } = render(
      <UserContext.Provider value={{ user: agent }}>
        <ScheduleAppointment
          workOrder={workOrder}
          hasExistingAppointment={false}
          workOrderReference={workOrder.reference}
        />
      </UserContext.Provider>
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('Shows an error state', () => {
    useDrsAppointmentScheduler.mockReturnValue({
      schedulerSessionId: null,
      isLoading: false,
      error: 'Some error occured',
      handleExternalLinkOpen: jest.fn(),
    })

    let workOrder = new WorkOrder(workOrderData)

    const { asFragment } = render(
      <UserContext.Provider value={{ user: agent }}>
        <ScheduleAppointment
          workOrder={workOrder}
          hasExistingAppointment={false}
          workOrderReference={workOrder.reference}
        />
      </UserContext.Provider>
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('Shows a success state', () => {
    useDrsAppointmentScheduler.mockReturnValue({
      schedulerSessionId: 'schedulerSessionId',
      isLoading: false,
      error: null,
      handleExternalLinkOpen: jest.fn(),
    })

    let workOrder = new WorkOrder(workOrderData)

    const { asFragment } = render(
      <UserContext.Provider value={{ user: agent }}>
        <ScheduleAppointment
          workOrder={workOrder}
          hasExistingAppointment={false}
          workOrderReference={workOrder.reference}
        />
      </UserContext.Provider>
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
