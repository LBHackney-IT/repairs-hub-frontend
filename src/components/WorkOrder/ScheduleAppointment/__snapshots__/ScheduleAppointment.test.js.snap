// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppointmentDetails component Shows a loading state 1`] = `
<DocumentFragment>
  <span
    style="color: rgb(100, 116, 139); font-size: 12px; display: flex; align-items: center; justify-content: flex-start;"
  >
    <svg
      data-testid="spinner-undefined"
      height="20"
      stroke="#00703c"
      viewBox="0 0 42 42"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        fill="none"
        fill-rule="evenodd"
      >
        <g
          stroke-width="5"
          transform="translate(3 3)"
        >
          <circle
            cx="18"
            cy="18"
            r="18"
            stroke-opacity=".5"
          />
          <path
            d="M36 18c0-9.94-8.06-18-18-18"
            transform="rotate(112.708 18 18)"
          >
            <animatetransform
              attributeName="transform"
              dur="1s"
              from="0 18 18"
              repeatCount="indefinite"
              to="360 18 18"
              type="rotate"
            />
          </path>
        </g>
      </g>
    </svg>
    <span
      style="margin: 0px 0px 0px 10px;"
    >
      Fetching DRS session...
    </span>
  </span>
</DocumentFragment>
`;

exports[`AppointmentDetails component Shows a success state 1`] = `
<DocumentFragment>
  <a
    class="lbh-link"
    href="undefined&sessionId=schedulerSessionId"
    rel="noopener"
    target="_blank"
  >
    <strong>
      Open DRS
    </strong>
     to book an appointment
  </a>
</DocumentFragment>
`;

exports[`AppointmentDetails component Shows an error state 1`] = `
<DocumentFragment>
  <span
    class="govuk-error-message lbh-error-message"
    data-error-id=""
    data-testid="error-message"
  >
    <span
      class="govuk-visually-hidden"
    >
      Error:
    </span>
     Some error occured
  </span>
</DocumentFragment>
`;
